explorer/interpreter/type_checker.cpp

```diff
@@ -3357,13 +3379,37 @@ auto TypeChecker::DeclareClassDeclaration(Nonnull<ClassDeclaration*> class_decl,
   ImplScope class_scope;
   class_scope.AddParent(scope_info.innermost_scope);
 
-  if (class_decl->extensibility() != ClassExtensibility::None) {
-    return ProgramError(class_decl->source_loc())
-           << "Class prefixes `base` and `abstract` are not supported yet";
-  }
-  if (class_decl->extends()) {
+  if (class_decl->extensibility() == ClassExtensibility::Abstract) {
     return ProgramError(class_decl->source_loc())
-           << "Class extension with `extends` is not supported yet";
+           << "Class prefix `abstract` is not supported yet";
+  }
+
+  std::optional<Nonnull<const NominalClassType*>> base_class;
+  if (class_decl->base().has_value()) {
+    Nonnull<Expression*> base_class_expr = *class_decl->base();
+    CARBON_RETURN_IF_ERROR(TypeCheckExp(base_class_expr, class_scope));
+    const auto& base_type = base_class_expr->static_type();
+    switch (base_type.kind()) {
+      case Value::Kind::TypeOfClassType:
+        base_class = &cast<TypeOfClassType>(base_type).class_type();
+        if (base_class.value()->declaration().extensibility() ==
+            ClassExtensibility::None) {
+          return ProgramError(class_decl->source_loc())
+                 << "Base class `" << base_class.value()->declaration().name()
+                 << "` is `final` and cannot inherited. Add the `base` or "
+                    "`abstract` class prefix to `"
+                 << base_class.value()->declaration().name()
```

Comment:
```suggestion
                 << base_type
```

I think the Value will yield a cleaner result here

Reply:
This (unfortunately) prints `typeof(class C)` instead of `C`. Would have been cleaner!